### PR TITLE
DD-96

### PR DESCRIPTION
**Pull request by lana-shanghai**
Add pay_as_bid_matching_algorithm

### CODE REVIEW

File: src/pay_as_bid.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                        already_selected_bids.push(bid.id);
+                        let selected_energy = bid.energy.min(offer.energy);
+                        let new_bid_offer_match = BidOfferMatch {
+                            market_id: market_id,
+                            bids: vec![bid],
+                            selected_energy: selected_energy,
+                            offers: vec![offer],
+                            trade_rate: bid.energy_rate,
+                        };
+                        bid_offer_pairs.push(new_bid_offer_match);
```
Comment (spyrostz) [inline src/pay_as_bid.rs:48]: I think you also need to return the bid_offer_pairs at the end of the method

File: src/pay_as_bid.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    bids: Vec<Bid>,
+    offers: Vec<Offer>,
+}
+
+pub trait GetMatchesRecommendations {
+    fn get_matches_recommations(&self, matching_data: MatchingData) {
+        let mut bid_offer_pairs = Vec::new();
+        for (market_id, data) in matching_data.iter().enumerate() {
+            let mut sorted_offers = matching_data.bids.sort_by(|a, b| b.energy_rate.cmp(&a.energy_rate));
+            let mut sorted_bids = matching_data.offers.sort_by(|a, b| b.energy_rate.cmp(&a.energy_rate));
```
[the remark this answers is not in the document]
Comment (lana-shanghai) [reply src/pay_as_bid.rs:30]: @faizan2590 good idea

File: src/pay_as_bid.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    energy_rate: f64,
+    energy: f64,
+    id: i8,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct MatchingData {
+    bids: Vec<Bid>,
+    offers: Vec<Offer>,
+}
```
[the remark this answers is not in the document]
Comment (lana-shanghai) [reply src/pay_as_bid.rs:23]: @faizan2590 Yes, was planning to do that!

File: src/pay_as_bid.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
+const FLOATING_POINT_TOLERANCE: f64 = 0.00001;
```
[the remark this answers is not in the document]
Comment (lana-shanghai) [reply src/pay_as_bid.rs:1]: @faizan2590 Agree

**Pull request merged**